subsys/net/lib/wifi_provision/wifi_provision.c

```diff
@@ -0,0 +1,995 @@
+		return 0;
+	}
+
+	/* Provision self-signed server certificate */
+	ret = tls_credential_add(CONFIG_WIFI_PROVISION_CERTIFICATE_SEC_TAG,
+				 TLS_CREDENTIAL_CA_CERTIFICATE,
+				 server_certificate,
+				 sizeof(server_certificate));
+	if (ret == -EEXIST) {
+		LOG_DBG("CA already exists, sec tag: %d",
+			CONFIG_WIFI_PROVISION_CERTIFICATE_SEC_TAG);
+	} else if (ret < 0) {
+		LOG_ERR("Failed to register CA certificate: %d", ret);
+		return ret;
+	}
+
+	ret = tls_credential_add(CONFIG_WIFI_PROVISION_CERTIFICATE_SEC_TAG,
+				 TLS_CREDENTIAL_SERVER_CERTIFICATE,
+				 server_certificate,
+				 sizeof(server_certificate));
+	if (ret == -EEXIST) {
+		LOG_DBG("Public certificate already exists, sec tag: %d",
+			CONFIG_WIFI_PROVISION_CERTIFICATE_SEC_TAG);
+	} else if (ret < 0) {
+		LOG_ERR("Failed to register public certificate: %d", ret);
+		return ret;
+	}
+
+	ret = tls_credential_add(CONFIG_WIFI_PROVISION_CERTIFICATE_SEC_TAG,
+				 TLS_CREDENTIAL_PRIVATE_KEY,
+				 server_private_key, sizeof(server_private_key));
+
+	if (ret == -EEXIST) {
+		LOG_DBG("Private key already exists, sec tag: %d",
+			CONFIG_WIFI_PROVISION_CERTIFICATE_SEC_TAG);
+	} else if (ret < 0) {
+		LOG_ERR("Failed to register private key: %d", ret);
+		return ret;
+	}
+
+	LOG_DBG("Self-signed server certificate provisioned");
+
+	net_mgmt_init_event_callback(&net_l2_mgmt_cb,
+				     net_mgmt_wifi_event_handler,
+				     NET_MGMT_WIFI);
+
+	net_mgmt_add_event_callback(&net_l2_mgmt_cb);
+
+	smf_set_initial(SMF_CTX(&state_object), &state[STATE_UNPROVISIONED]);
+
+	http_parser_settings_init(&parser_settings);
+
+	parser_settings.on_body = on_body;
+	parser_settings.on_headers_complete = on_headers_complete;
+	parser_settings.on_message_begin = on_message_begin;
+	parser_settings.on_message_complete = on_message_complete;
+	parser_settings.on_url = on_url;
+
+	/* Start processing incoming IPv4 clients */
+	ret = process_tcp4();
+	if (ret < 0) {
+		LOG_ERR("Failed to start TCP server %d", ret);
+		return ret;
+	}
+
+	k_sem_take(&wifi_provision_sem, K_FOREVER);
+	return 0;
+}
+
+int wifi_provision_reset(void)
+{
+	LOG_DBG("Resetting Wi-Fi provision state machine");
+	new_event(EVENT_RESET);
+	return 0;
+}
+
+/* Function used to schedule state transitions. State transitions are offloaded to a separate
+ * workqueue to avoid races with Zephyr NET management. Time consuming tasks can safely be offloaded
+ * to this workqueue.
+ */
+static void wifi_provision_task(void)
+{
+	int ret;
+	enum module_event event_new = 0;
+
+	while (1) {
+
+		ret = k_msgq_get(&wifi_provision_msgq, &event_new, K_FOREVER);
+		if (ret) {
+			LOG_ERR("k_msgq_get, error: %d", ret);
+			notify_app(WIFI_PROVISION_EVT_FATAL_ERROR);
+			return;
+		}
+
+		state_object.event_next = event_new;
+
+		ret = smf_run_state(SMF_CTX(&state_object));
+		if (ret) {
+			LOG_ERR("smf_run_state, error: %d", ret);
+			notify_app(WIFI_PROVISION_EVT_FATAL_ERROR);
+			return;
+		}
+	}
+}
+
+K_THREAD_DEFINE(wifi_provision_task_id, 8192,
```

Comment:
This stack seems quite heavy. Though personally, I'd reconsider if we need a state machine at all, it seems that all those actions (scan (if needed), AP enable,  DHCP enable, HTTP server processing) could just take place sequentially, w/o a fancy state machine behind it. Separate thread here is IMO expensive choice.

Reply:
A state machine is always nice. Making the library more robust and easier to maintain. Things tend to get messy without some sort of structure keeping order. You also protect against out-of-order behavior without having flags everywhere to check if some functionality is triggered when it shouldn't.

I guess an argument can be made for that _nothing else should happen when provisioning is ongoing_ but I don't want to make that assumption.

I initially did something much simpler but experience a lot of issues that were typically solved by scheduling actions in a separate wq. Instead of calling the respective functionality in the net mgmt handlers. Not sure if it was a prio thing or not.

With the additional wq I found the overhead of the current structure acceptable as memory wise it was not far away from what I already had.

I'm ofc open to optimizing the library if it doesn't impose too many drawbacks in terms of stability and maintainability.